module/reports/helper.js

```diff
@@ -0,0 +1,443 @@
+                     let today = moment();
+                     startFrom = today.startOf('week').format('YYYY-MM-DD');
+                     endOf = today.endOf('week').format('YYYY-MM-DD');
+                } else if (reportType == 1) {
+                    endOf = moment().subtract(0, 'months').endOf('month').format('YYYY-MM-DD');
+                    startFrom = moment().subtract(0, 'months').startOf('month').format('YYYY-MM-DD');
+                } else if (reportType == 3) {
+                     startFrom = moment().quarter(moment().quarter()).startOf('quarter').format('YYYY-MM-DD');
+                     endOf = moment().quarter(moment().quarter()).endOf('quarter').format('YYYY-MM-DD');
+                }
+
+                query['$or'] = [
+                    { "lastSync" : { $gte: new Date(startFrom), $lte: new Date(endOf) } },
+                    { "tasks" : { $elemMatch: { lastSync: { $gte: new Date(startFrom), $lte: new Date(endOf) } } } },
+                ]
+
+                if (programId) {
+                    query['programInformation._id'] = mongoose.Types.ObjectId(programId);
+                }
+
+                const projectDetails = await UserProjectsHelper.projectDocument(
+                    query,
+                    ["taskReport", "status", "tasks", "categories"],
+                    []
+                );
+
+                let tasksReport = {
+                    "total":0,
+                    "completed": 0,
+                    "inProgress": 0,
+                    "notStarted": 0,
+                    "overdue":0
+                };
+                let categories = {
+                    "total":0
+                };
+                let categoriesArray = [];
+                let projectReport = {
+                    "total": 0,
+                    "completed": 0,
+                    "inProgress": 0,
+                    "notStarted": 0,
+                    "overdue": 0,
+                };
+
+                if (!projectDetails.length > 0) {
+
+                    return resolve({
+                        message: CONSTANTS.apiResponses.REPORTS_DATA_NOT_FOUND,
+                        result:  {
+                            data_available:false,
+                            data :{ 
+                                categories: categories,
+                                tasks: tasksReport,
+                                projects: projectReport
+                            }
+                        }
+                    })
+                }
+
+                projectReport["total"] =projectDetails.length;
+
+                   
+                await Promise.all(projectDetails.map(async function (project) {
+
+                    if (project.categories) {
+                        project.categories.map(category => {
+                            if (categoriesArray.includes(category.name)) {
+                                categories[category.name] = categories[category.name] + 1;
```

Comment:
@rakeshSgr - Let's not use the name as keys, it can have spaces special characters or different language

Reply:
resolved